feat: add express session

nictownsend opened the pull request:
- in memory session
- cookie keyed by server name

Contributes to: strimzi/strimzi-ui#107

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

nictownsend committed:
feat: add express session

- in memory session
- cookie keyed by server name
- default values for security (to be later updated)

Contributes to: #107

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

JQrdan commented on server/core/app.ts:
```diff
@@ -26,6 +27,16 @@ export const returnExpress: (
+    secret: 'CHANGEME',
+    name: serverName,
+    cookie: {
+      maxAge: 1000 * 3600 * 24 * 30,
```
Could you add a comment to say how long this is

JQrdan commented on server/core/app.ts:
```diff
@@ -26,6 +27,16 @@ export const returnExpress: (
   app.use(generateHttpLogger());
 
+  //Add session middleware
+  const sessionOpts: session.SessionOptions = {
```
Would be neater to import SessionOptions independently

JQrdan reviewed (changes requested):
Just a couple of minor things but otherwise LGTM

nictownsend committed:
chore: review comments from @JQrdan

- import session types directly
- comments clarifying session options

Contributes to: strimzi/strimzi-ui#107

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

matthew-chirgwin commented on server/core/app.ts:
```diff
@@ -26,6 +27,16 @@ export const returnExpress: (
   app.use(generateHttpLogger());
 
+  //Add session middleware
+  const sessionOpts: SessionOptions = {
```
I assume (in later issues) this will be enabled/added only when needed?

nictownsend replied:
I think it's fair to always have a session as we might want to store information for the lifetime of a browser session. In particular, for logging - @ajborley, would it be useful to store a "client id" in the session?

matthew-chirgwin commented on server/core/core.steps.ts:
```diff
@@ -37,90 +37,65 @@ When(
-      (err) => {
-        throw err;
-      }
+    return Object.entries(headerExpectedValue).reduce(
```
This is a lot more subtle/less readable than the prior version IMO. Can a comment be added to explain what this is doing/how the `chainedRequest` is being invoked?

nictownsend replied:
I'll revert - if it needs a comment then the efficiency isn't worth the readable impact

matthew-chirgwin commented on server/test_common/commonServerSteps.ts:
```diff
@@ -67,7 +69,22 @@ And(
-      server: request(returnExpress('test-server', () => world.configuration)),
+      server: request(
+        returnExpress(
+          world.serverName || 'test-server',
```
WIth the removal of the name argument this will need to be updated

nictownsend replied:
Just spotted it :'(

matthew-chirgwin reviewed (changes requested):
LGTM - just one test needs a comment for me to (when we return to this in future) we can see how the assertions are being chained

nictownsend committed:
fix: retrieve session name from config

- add session section to config
- default session name if not in config

Contributes to: strimzi/strimzi-ui#107

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

matthew-chirgwin commented on config/static.ts:
```diff
@@ -36,6 +36,9 @@ const server: Config<Literal> = {
         port: 9080,
         transport: {},
       },
+      session: {
```
no issue with the being here - can the typing and doc be updated to reflect? https://github.com/matthew-chirgwin/strimzi-ui/tree/master/server#configuration-options / https://github.com/matthew-chirgwin/strimzi-ui/blob/master/server/types.ts

nictownsend replied:
Adding

matthew-chirgwin reviewed (commented):
LGTM - just one comment re typing/doc

nictownsend committed:
chore: review comments from @matthew-chirgwin

- update configuration docs

Contributes to: strimzi/strimzi-ui#107

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

matthew-chirgwin commented on server/types.ts:
```diff
@@ -79,6 +83,7 @@ export type serverConfig = {
   hostname: string;
   /** The port number to use/accept traffic on */
   port: number;
+  session: sessionConfigType;
```
JSDoc for usage/context/purpose?
